### DISCUSSION TITLE
Is shadowing-in-nested-functions C# too free that people can reference outside var and define an new var with the same name in same method without any warn

### DISCUSSION THREAD

**Discussion by yyjdelete** [dotnet/csharplang#5327] [General] [answered]
Is shadowing-in-nested-functions C# too free that people can reference outside var and define an new one with the same name in one method without any warn

See the below code. Maybe it's difficult to debug and recognize there are 2 different var with the same name `val` in `Test2`

```cs
using System;

public class Class1
{
    public void Test1(int val)
    {
        void Test2()
        {
            Console.WriteLine(val);//use var from outside scope
            var array = new int[]{1};
            foreach (var val in array)
            {
                Console.WriteLine(val);//use var from current scope
            }
            Console.WriteLine(val);//still use var from outside scope
        }
    }
}
```

**Comment by 333fred** [comment]
We don't think so.

**Reply by jnm2** [reply]
Additionally, a community analyzer could be created which warns when this happens.

**Comment by 333fred** [comment, marked as answer]
This was discussed in LDM [on 11/03/2021](https://github.com/dotnet/csharplang/blob/main/meetings/2021/LDM-2021-11-03.md#name-shadowing-in-local-functions). TLDR: we somewhat regret that you can both shadow and capture in the same local function, but we think it's not worth a warning wave or language change to correct at this point.
